File: CPP/Data_Structures/CircularLinkedList.cpp
#include <iostream>
using namespace std;
struct node
{
    int value;
    node *next;
};

node *createNode(int value)
{
    node *n1 = new node;
    n1->value = value;
    n1->next = NULL;
    return n1;
}

class CircularLinkedList
{
    node *last;

public:
    CircularLinkedList() { last = NULL; }
    void addToEmpty(int value)
    {
        node *newNode = createNode(value);
        newNode->next = newNode;
        last = newNode;
    }
    void insertBeginning(int value)
    {
        if (last == NULL)
        {
            addToEmpty(value);
            return;
        }
        node *newNode = createNode(value);
        newNode->next = last->next;
        last->next = newNode;
    }
    void insertEnd(int value)
    {
        if (last == NULL)
        {
            addToEmpty(value);
            return;
        }
        node *newNode = createNode(value);
        newNode->next = last->next;
        last->next = newNode;
        last = newNode;
    }
    void deleteBeginning()
    {
        if (last == NULL)
        {
            cout << "Underflow\n";
            return;
        }
        node *ptr = last->next;
        last->next = ptr->next;
        free(ptr);
    }
    void displayAll()
    {
        if (last == NULL)
        {
            cout << "Underflow\n";
            return;
        }
        node *ptr = last->next;
        while (true)
        {
            cout << ptr->value << ' ';
            ptr = ptr->next;
            if (ptr == last)
            {
                cout << ptr->value;
                break;
            }
        }
        cout << endl;
    }
};

// Test Code

int main()
{
    CircularLinkedList CLL;
    CLL.insertBeginning(1);
    CLL.insertBeginning(2);
    CLL.insertBeginning(3);
    CLL.displayAll();
    CLL.insertEnd(4);
    CLL.displayAll();
    CLL.deleteBeginning();
    CLL.displayAll();
    return 0;
}

File: Python/Data_Structures/CircularLinkedList.py
class Node(object):
    def __init__(self, value):
        self.value = value
        self.next = None


class CircularLinkedList(object):
    def __init__(self):
        self.last = None

    def addToEmpty(self, value):
        newNode = Node(value)
        self.last = newNode
        self.last.next = newNode

    def insertBeginning(self, value):
        if self.last is None:
            self.addToEmpty(value)
            return
        newNode = Node(value)
        newNode.next = self.last.next
        self.last.next = newNode

    def insertEnd(self, value):
        if self.last is None:
            self.addToEmpty(value)
            return
        newNode = Node(value)
        newNode.next = self.last.next
        self.last.next = newNode
        self.last = newNode

    def deleteBeginning(self):
        if self.last is None:
            raise ValueError("Underflow")
        ptr = self.last.next
        self.last.next = ptr.next
        del ptr

    def displayAll(self):
        ptr = self.last.next
        while True:
            print(ptr.value, end=' ')
            ptr = ptr.next
            if ptr is self.last:
                print(ptr.value, end=' ')
                break
        print()

# Test Code


CLL = CircularLinkedList()
CLL.insertBeginning(1)
CLL.insertBeginning(2)
CLL.insertBeginning(3)
CLL.displayAll()
CLL.insertEnd(4)
CLL.displayAll()
CLL.deleteBeginning()
CLL.displayAll()
